src/Symfony/Component/Form/Extension/Core/Type/CollectionType.php

```diff
@@ -81,24 +73,78 @@ public function finishView(FormView $view, FormInterface $form, array $options)
      */
     public function configureOptions(OptionsResolver $resolver)
     {
-        $entryOptionsNormalizer = function (Options $options, $value) {
-            $value['block_name'] = 'entry';
+        $entryOptionsNormalizer = function (Options $options, $entryOptions) {
+            if (is_array($entryOptions)) {
+                $entryOptions['block_name'] = 'entry';
+
+                return $entryOptions;
+            }
+
+            // Else it's a callable, just wrap it.
+            return function ($entry) use ($entryOptions) {
+                $dynamicOptions = call_user_func($entryOptions, $entry);
+
+                if (!is_array($dynamicOptions)) {
+                    throw new InvalidOptionsException(sprintf('The "entry_options" callable should return an array but got %s', is_object($dynamicOptions) ? get_class($dynamicOptions) : gettype($dynamicOptions)));
+                }
+
+                $dynamicOptions['block_name'] = 'entry';
 
-            return $value;
+                return $dynamicOptions;
+            };
+        };
+
+        $prototypeOptionsNormalizer = function (Options $options, $value) {
+            $entryOptions = is_array($options['entry_options'])
+                ? $options['entry_options']
```

Comment:
Please use a ``if`` rather than a ternary. Nesting 3 ternary operators is totally unreadable